tests/browser_context_test.go

```diff
@@ -750,119 +745,50 @@ func TestBrowserContextWaitForEvent(t *testing.T) {
-			// If we run them synchronously then waitForEvent will wait
-			// indefinitely and eventually the test will timeout.
-			err = tb.run(
-				ctx,
-				func() error {
-					op := optsOrPredicateToGojaValue(t, tb, tc.optsOrPredicate)
-					resp, err := bc.WaitForEvent(tc.event, op)
-					if resp != nil {
-						var ok bool
-						p1, ok = resp.(*common.Page)
-						require.True(t, ok)
-					}
-					return err
-				},
-				func() error {
-					var err error
-					p2, err = bc.NewPage()
-					return err
-				},
+			var (
+				aboutToCallWait = make(chan bool)
+				waitDone        = make(chan bool)
+				p1ID, p2ID      string
 			)
 
-			if tc.wantErr == "" {
-				assert.NoError(t, err)
-				// We want to make sure that the page that was created with
-				// newPage matches the return value from waitForEvent.
-				assert.Equal(t, p1.MainFrame().ID(), p2.MainFrame().ID())
-				return
-			}
-
-			assert.ErrorContains(t, err, tc.wantErr)
-		})
-	}
-}
+			go func() {
+				defer close(waitDone)
 
-// optsOrPredicate is a helper type to enable us to package up the optional
-// arguments which could either be a predicate function, or the options object
-// which contains a predicate function and a timeout.
-type optsOrPredicate struct {
-	predicate     *string
-	timeout       *int64
-	justPredicate *string
-}
-
-func stringPtr(value string) *string {
-	strPointer := new(string)
-	*strPointer = value
-	return strPointer
-}
-
-func int64Ptr(value int64) *int64 {
-	int64Pointer := new(int64)
-	*int64Pointer = value
-	return int64Pointer
-}
+				var resp any
+				close(aboutToCallWait)
+				resp, err = bc.WaitForEvent(tc.event, tc.predicate, tc.timeout)
+				if resp != nil {
+					p, ok := resp.(*common.Page)
+					require.True(t, ok)
 
-// optsOrPredicateToGojaValue will take optsOrPredicate and correctly define the
-// optional options where necessary. It will either return nil, a predicate
-// function or an options object.
-func optsOrPredicateToGojaValue(t *testing.T, tb *testBrowser, op *optsOrPredicate) goja.Value {
-	t.Helper()
+					p1ID = p.MainFrame().ID()
+				}
+			}()
 
-	// Options or predicate are undefined.
-	if op == nil {
-		return nil
-	}
+			<-aboutToCallWait
 
-	// The optional argument is a predicate function.
-	if op.justPredicate != nil {
-		predicate, err := tb.runJavaScript(*op.justPredicate)
-		require.NoError(t, err)
-		return predicate
-	}
+			if tc.wantErr == "" {
+				p, err := bc.NewPage()
+				require.NoError(t, err)
 
-	// The option argument is a options object with only the predicate function
-	// defined but no timeout.
-	if op.predicate != nil && op.timeout == nil {
-		predicate, err := tb.runJavaScript(*op.predicate)
-		require.NoError(t, err)
+				p2ID = p.MainFrame().ID()
+			}
 
-		opts := tb.toGojaValue(struct {
-			Predicate goja.Value
-		}{
-			Predicate: predicate,
-		})
+			select {
+			case <-waitDone:
+			case <-ctx.Done():
+				err = ctx.Err()
```

Comment:
Should we assert here that this error is `nil` which would mean that the context was not canceled due to TO? Or maybe use another error, so we don't overlap `err` which is used to store the error resulting from `bc.WaitForEvent` and assert it separately?

Reply:
I see where you are getting at, i'm wondering whether it might be clearer to assert the context canceled error separately.

The idea is that the `err` is checked in the lines below. If `wantErr` is an empty string, then `err` should be `nil`. If we expect an error then it should match `wantErr`.

Reply:
Yes, I understood how it works, but it was a little bit tricky. That's why I thought it could make sense to be more explicit on asserting both possible errors. But no strong opinion, we can leave it as it is also.

Reply:
Sure, i've added an assert to make it clearer: e18558adba1a3cbef4eaa6b05abdf36323f1f420